pkg/controller/gitsource/gitsource_controller.go

```diff
@@ -0,0 +1,139 @@
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("gitsource-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource GitSource
+	err = c.Watch(&source.Kind{Type: &v1alpha1.GitSource{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileGitSource{}
+
+// ReconcileGitSource reconciles a GitSource object
+type ReconcileGitSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a GitSource object and makes changes based on the state read
+// and what is in the GitSource.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileGitSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling GitSource")
+
+	// Fetch the GitSource instance
+	gitSource := &v1alpha1.GitSource{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, gitSource)
+	if err != nil {
+		reqLogger.Error(err, "Error getting GitSource object")
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	gitSourceLogger := gslog.LogWithGSValues(reqLogger, gitSource)
+
+	isDirty := updateStatus(gitSourceLogger, gitSource)
+
+	if isDirty {
+		err = r.client.Update(context.TODO(), gitSource)
+		if err != nil {
+			gitSourceLogger.Error(err, "Error updating GitSource object")
+			if errors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+				// Return and don't requeue
+				return reconcile.Result{}, nil
+			}
+			// Error updating the object - requeue the request.
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func updateStatus(log *gslog.GitSourceLogger, gitSource *v1alpha1.GitSource) (isDirty bool) {
+
+	if gitSource.Status.Connection.State != "" {
+		return false
+	}
+	if gitSource.Status.State == "" {
+		gitSource.Status.State = v1alpha1.Initializing
```

Comment:
@MatousJobanek , this doesn't move to any other state, we should probably move this to `initialized` ?

Reply:
The idea behind this is that the state should be either `ready` or `initialized` where `initialized` is the first step, then you can move it to `ready`. If the state is empty - the CR was just created - then we could assume that the CR is in `initialized` state - that is the reason why I set it here.